snuba/clickhouse/columns.py

```diff
@@ -1,21 +1,82 @@
 from __future__ import annotations
 
+from abc import ABC, abstractmethod
 from itertools import chain
 from typing import (
-    cast,
     Iterator,
+    List,
     Mapping,
     MutableMapping,
-    List,
     Optional,
     Sequence,
     Tuple,
+    Type,
     Union,
+    cast,
 )
 
 from snuba.clickhouse.escaping import escape_identifier
 
 
+class TypeModifier(ABC):
+    """
+    Changes the semantics of a ColumnType. Usual modifiers are
+    Nullable, WithDefault, or similar.
+    """
+
+    @abstractmethod
+    def for_schema(self, content: str) -> str:
+        """
+        Formats the modifier for DDL statements.
+        The content parameter is the serialized type the modifier
+        applies to.
+        """
+        raise NotImplementedError
+
+
+class ColumnType:
+    def __init__(self, modifiers: Optional[List[TypeModifier]] = None):
+        """
+        The modifiers list to apply to this type. Modifiers are applied
+        to the type in the same order they appear in the list.
+        `String([LowCardinality(), Nullable()])`
+        means
+        `LowCardinality(Nullable(String))`
+        """
+        self.__modifiers = modifiers or []
+
+    def __repr__(self) -> str:
```

Comment:
Should this and `__eq__` include the modifiers as well?

Reply:
Yeah, it appears that we are no longer comparing the columns + modifiers like we used to

Reply:
Mh, this is weird. If this was not comparing modifiers, a test around parsing the schema should fail. Let me check what is going on there.